File: bl/synthesizer.py
"""
bl/synthesizer.py — Campaign synthesizer.

Reads all findings/*.md + results.tsv after each wave and calls Claude
to produce synthesis.md: validated bets, dead ends, unvalidated bets,
and a CONTINUE|STOP|PIVOT recommendation.
"""

import shutil
import subprocess
import sys
from pathlib import Path


_MAX_CORPUS_CHARS = 12000


def _build_findings_corpus(findings_dir: Path, results_tsv: Path) -> str:
    """
    Reads all findings/*.md files (sorted by name) and results.tsv.
    Returns a combined text corpus for the LLM prompt.
    Cap total size at 12000 chars (truncate oldest findings first if over limit).
    """
    # Results TSV first (always include)
    tsv_section = ""
    if results_tsv.exists():
        tsv_content = results_tsv.read_text(encoding="utf-8").strip()
        if tsv_content:
            tsv_section = f"## Results Summary\n\n{tsv_content}\n"

    # Findings files sorted by name
    finding_sections: list[str] = []
    if findings_dir.exists():
        for finding_file in sorted(findings_dir.glob("*.md")):
            content = finding_file.read_text(encoding="utf-8").strip()
            if content:
                finding_sections.append(f"### {finding_file.stem}\n\n{content}\n")

    # Build corpus: TSV always included, truncate low-severity findings first if over limit
    tsv_chars = len(tsv_section)
    budget = _MAX_CORPUS_CHARS - tsv_chars

    # F11.3: severity-aware truncation — drop COMPLIANT/FIXED findings before FAILURE/NON_COMPLIANT
    # Sort: high-severity (FAILURE, NON_COMPLIANT) first, low-severity (COMPLIANT, FIXED) last
    _HIGH_SEVERITY = frozenset(
        {
            "FAILURE",
            "NON_COMPLIANT",
            "WARNING",
            "REGRESSION",
            "ALERT",
            "DIAGNOSIS_COMPLETE",
            "FIX_FAILED",
        }
    )

    def _finding_priority(section: str) -> int:
        """Lower number = higher priority = kept longer under budget pressure."""
        for verdict in _HIGH_SEVERITY:
            if f"**Verdict**: {verdict}" in section or f": {verdict}" in section:
                return 0  # high severity — keep
        return 1  # low severity (COMPLIANT, FIXED, etc.) — drop first

    finding_sections.sort(key=_finding_priority)  # high-severity first
    while finding_sections and budget < sum(len(s) for s in finding_sections):
        finding_sections.pop()  # drop from tail (low-severity last)

    corpus_parts = []
    if tsv_section:
        corpus_parts.append(tsv_section)
    if finding_sections:
        corpus_parts.append("## Findings\n\n" + "\n".join(finding_sections))

    if not corpus_parts:
        return "No findings or results available yet."

    return "\n".join(corpus_parts)


def _read_doctrine(project_dir: Path) -> str:
    """
    Reads doctrine.md from project_dir if it exists.
    Returns content, or empty string if not found.
    """
    doctrine_path = project_dir / "doctrine.md"
    if not doctrine_path.exists():
        return ""
    return doctrine_path.read_text(encoding="utf-8").strip()


def _call_claude(prompt: str) -> str | None:
    """
    Calls Claude via subprocess: claude -p "{prompt}" --output-format text
    Returns stdout stripped, or None on failure.
    """
    claude_bin = shutil.which("claude") or "claude"
    try:
        result = subprocess.run(
            [claude_bin, "-p", prompt, "--output-format", "text"],
            capture_output=True,
            text=True,
            encoding="utf-8",
            timeout=120,
        )
        if result.stderr.strip():
            print(result.stderr.strip(), file=sys.stderr)
        if result.returncode != 0:
            print(
                f"[synthesizer] Claude exited with code {result.returncode}",
                file=sys.stderr,
            )
            return None
        return result.stdout.strip() or None
    except FileNotFoundError:
        print("[synthesizer] claude CLI not found", file=sys.stderr)
        return None
    except subprocess.TimeoutExpired:
        print("[synthesizer] Claude timed out after 120s", file=sys.stderr)
        return None
    except Exception as e:
        print(f"[synthesizer] Unexpected error calling Claude: {e}", file=sys.stderr)
        return None


def parse_recommendation(synthesis_text: str) -> str:
    """
    Extracts the recommendation from synthesis output.
    Scans lines after the '## Recommended Next Action' section header.
    Falls back to full-text scan if the section header is absent.
    Returns "CONTINUE" as default if not found.
    """
    lines = synthesis_text.splitlines()

    # F10.2: locate the Recommended Next Action section, scan only lines after it
    section_start = None
    for i, line in enumerate(lines):
        if "recommended next action" in line.lower():
            section_start = i + 1
            break

    scan_lines = lines[section_start:] if section_start is not None else lines

    for line in scan_lines:
        upper = line.upper()
        if "STOP" in upper:
            return "STOP"
        if "PIVOT" in upper:
            return "PIVOT"
        if "CONTINUE" in upper:
            return "CONTINUE"
    return "CONTINUE"


def synthesize(
    project_dir: Path,
    wave: int | None = None,
    dry_run: bool = False,
) -> Path | None:
    """
    Main entry point. Builds corpus, reads doctrine, calls Claude,
    writes synthesis.md (or prints on dry_run).

    Returns path to synthesis.md, or None on dry_run or failure.
    """
    findings_dir = project_dir / "findings"
    results_tsv = project_dir / "results.tsv"

    corpus = _build_findings_corpus(findings_dir, results_tsv)
    doctrine = _read_doctrine(project_dir)

    # Detect wave from corpus if not supplied
    wave_label = wave if wave is not None else "?"

    doctrine_section = ""
    if doctrine:
        doctrine_section = f"## Project Doctrine\n\n{doctrine}\n\n"

    prompt = f"""You are a research campaign director reviewing findings from a BrickLayer autonomous research campaign.

Your job: synthesize the accumulated evidence and produce a structured campaign status report.

{doctrine_section}CAMPAIGN FINDINGS:
{corpus}

Produce a synthesis report in this EXACT format:

# Campaign Synthesis — Wave {wave_label}

## Core Hypothesis Verdict
[CONFIRMED | UNCONFIRMED | PARTIALLY CONFIRMED | REFUTED] — one paragraph explaining why.

## Validated Bets
List each thing the campaign has confirmed with evidence. Format:
- [what was confirmed]: evidence from [Q IDs] — confidence [high/medium/low]

If nothing validated yet: "None confirmed yet."

## Dead Ends
List paths that have been exhausted with no findings worth pursuing further:
- [what was tested]: [Q IDs] found nothing actionable — stop probing here

If none: "None identified."

## Unvalidated Bets
List key questions or assumptions that have NOT been tested yet:
- [what is unvalidated]: [why it matters]

If all bets validated: "None remaining."

## Recommended Next Action
State exactly one of: CONTINUE, STOP, or PIVOT

CONTINUE — more questions needed, campaign is making progress
STOP — core hypothesis confirmed or refuted with sufficient confidence; no unvalidated bets remain
PIVOT — a new direction has emerged that the current question bank doesn't cover; describe the pivot

Then one paragraph of specific reasoning for the recommendation."""

    output = _call_claude(prompt)

    if output is None:
        print(
            "[synthesizer] Claude call failed — no synthesis written", file=sys.stderr
        )
        return None

    recommendation = parse_recommendation(output)
    print(f"[synthesizer] Recommendation: {recommendation}", file=sys.stderr)

    if dry_run:
        print(output)
        return None

    synthesis_path = project_dir / "synthesis.md"
    synthesis_path.write_text(output, encoding="utf-8")
    return synthesis_path

File: masonry/src/engine/synthesizer.js
'use strict';
// engine/synthesizer.js — Campaign synthesizer.
//
// Port of bl/synthesizer.py to Node.js. Reads findings + results.tsv after
// each wave and calls Claude CLI to produce synthesis.md.

const { execFileSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const MAX_CORPUS_CHARS = 12000;

const HIGH_SEVERITY = new Set([
  'FAILURE', 'NON_COMPLIANT', 'WARNING', 'REGRESSION',
  'ALERT', 'DIAGNOSIS_COMPLETE', 'FIX_FAILED',
]);

// ---------------------------------------------------------------------------
// Corpus builder
// ---------------------------------------------------------------------------

function _buildFindingsCorpus(findingsDir, resultsTsv) {
  let tsvSection = '';
  if (fs.existsSync(resultsTsv)) {
    const tsvContent = fs.readFileSync(resultsTsv, 'utf8').trim();
    if (tsvContent) tsvSection = `## Results Summary\n\n${tsvContent}\n`;
  }

  let findingSections = [];
  if (fs.existsSync(findingsDir)) {
    const files = fs.readdirSync(findingsDir).filter(f => f.endsWith('.md')).sort();
    for (const file of files) {
      const content = fs.readFileSync(path.join(findingsDir, file), 'utf8').trim();
      if (content) {
        findingSections.push(`### ${path.basename(file, '.md')}\n\n${content}\n`);
      }
    }
  }

  const tsvChars = tsvSection.length;
  const budget = MAX_CORPUS_CHARS - tsvChars;

  // Severity-aware truncation: high-severity first, drop low-severity tail
  findingSections.sort((a, b) => _findingPriority(a) - _findingPriority(b));
  while (findingSections.length && budget < findingSections.reduce((s, f) => s + f.length, 0)) {
    findingSections.pop();
  }

  const parts = [];
  if (tsvSection) parts.push(tsvSection);
  if (findingSections.length) parts.push('## Findings\n\n' + findingSections.join('\n'));

  return parts.length ? parts.join('\n') : 'No findings or results available yet.';
}

function _findingPriority(section) {
  for (const verdict of HIGH_SEVERITY) {
    if (section.includes(`**Verdict**: ${verdict}`) || section.includes(`: ${verdict}`)) {
      return 0;
    }
  }
  return 1;
}

// ---------------------------------------------------------------------------
// Doctrine
// ---------------------------------------------------------------------------

function _readDoctrine(projectDir) {
  const doctrinePath = path.join(projectDir, 'doctrine.md');
  if (!fs.existsSync(doctrinePath)) return '';
  return fs.readFileSync(doctrinePath, 'utf8').trim();
}

// ---------------------------------------------------------------------------
// Claude CLI
// ---------------------------------------------------------------------------

function _callClaude(prompt) {
  const claudeBin = _findClaude();
  try {
    const result = execFileSync(claudeBin, ['-p', prompt, '--output-format', 'text'], {
      encoding: 'utf8',
      timeout: 120000,
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    return result.trim() || null;
  } catch (err) {
    if (err.code === 'ENOENT') {
      process.stderr.write('[synthesizer] claude CLI not found\n');
    } else if (err.killed) {
      process.stderr.write('[synthesizer] Claude timed out after 120s\n');
    } else {
      process.stderr.write(`[synthesizer] Claude error: ${err.message}\n`);
    }
    return null;
  }
}

function _findClaude() {
  // Check common locations
  const { execSync } = require('child_process');
  try {
    return execSync('which claude', { encoding: 'utf8' }).trim();
  } catch (_) {
    return 'claude';
  }
}

// ---------------------------------------------------------------------------
// Recommendation parser
// ---------------------------------------------------------------------------

function parseRecommendation(synthesisText) {
  const lines = synthesisText.split('\n');

  let sectionStart = null;
  for (let i = 0; i < lines.length; i++) {
    if (lines[i].toLowerCase().includes('recommended next action')) {
      sectionStart = i + 1;
      break;
    }
  }

  const scanLines = sectionStart !== null ? lines.slice(sectionStart) : lines;

  for (const line of scanLines) {
    const upper = line.toUpperCase();
    if (upper.includes('STOP')) return 'STOP';
    if (upper.includes('PIVOT')) return 'PIVOT';
    if (upper.includes('CONTINUE')) return 'CONTINUE';
  }
  return 'CONTINUE';
}

// ---------------------------------------------------------------------------
// Retrospective
// ---------------------------------------------------------------------------

function _runRetrospective(projectDir) {
  const agentFile = path.join(projectDir, '.claude', 'agents', 'retrospective.md');
  if (!fs.existsSync(agentFile)) return;

  const questionsMd = path.join(projectDir, 'questions.md');
  let doneCount = 0;
  let totalCount = 0;
  if (fs.existsSync(questionsMd)) {
    const text = fs.readFileSync(questionsMd, 'utf8');
    totalCount = (text.match(/\*\*Status\*\*:/g) || []).length;
    doneCount = (text.match(/\*\*Status\*\*: DONE/g) || []).length;
  }

  let projectName = path.basename(projectDir);
  const briefPath = path.join(projectDir, 'project-brief.md');
  if (fs.existsSync(briefPath)) {
    const firstLine = fs.readFileSync(briefPath, 'utf8').split('\n')[0].replace(/^#\s*/, '').trim();
    if (firstLine) projectName = firstLine;
  }

  const prompt = `Act as the retrospective agent defined in ${agentFile}. ` +
    `Project: ${projectName}. Project directory: ${projectDir}. ` +
    `Campaign stats: ${totalCount} questions total, ${doneCount} DONE. ` +
    'Complete all three parts: (1) process scoring, (2) content integrity analysis, (3) LLM self-report. ' +
    'Write retrospective.md to the project root.';

  const claudeBin = _findClaude();
  try {
    execFileSync(claudeBin, ['-p', prompt, '--output-format', 'text'], {
      encoding: 'utf8',
      timeout: 300000,
      cwd: projectDir,
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    process.stderr.write('[synthesizer] Retrospective complete.\n');
  } catch (err) {
    process.stderr.write(`[synthesizer] Retrospective skipped: ${err.message}\n`);
  }
}

// ---------------------------------------------------------------------------
// Main API
// ---------------------------------------------------------------------------

function synthesize(projectDir, wave, dryRun = false) {
  projectDir = path.resolve(projectDir);
  const findingsDir = path.join(projectDir, 'findings');
  const resultsTsv = path.join(projectDir, 'results.tsv');

  const corpus = _buildFindingsCorpus(findingsDir, resultsTsv);
  const doctrine = _readDoctrine(projectDir);

  const waveLabel = wave != null ? wave : '?';

  const doctrineSection = doctrine ? `## Project Doctrine\n\n${doctrine}\n\n` : '';

  const prompt = `You are a research campaign director reviewing findings from a BrickLayer autonomous research campaign.

Your job: synthesize the accumulated evidence and produce a structured campaign status report.

${doctrineSection}CAMPAIGN FINDINGS:
${corpus}

Produce a synthesis report in this EXACT format:

# Campaign Synthesis — Wave ${waveLabel}

## Core Hypothesis Verdict
[CONFIRMED | UNCONFIRMED | PARTIALLY CONFIRMED | REFUTED] — one paragraph explaining why.

## Validated Bets
List each thing the campaign has confirmed with evidence.

## Dead Ends
List paths that have been exhausted.

## Unvalidated Bets
List key questions or assumptions not yet tested.

## Recommended Next Action
State exactly one of: CONTINUE, STOP, or PIVOT
Then one paragraph of specific reasoning.`;

  const output = _callClaude(prompt);
  if (!output) {
    process.stderr.write('[synthesizer] Claude call failed — no synthesis written\n');
    return null;
  }

  const recommendation = parseRecommendation(output);
  process.stderr.write(`[synthesizer] Recommendation: ${recommendation}\n`);

  if (dryRun) {
    process.stdout.write(output + '\n');
    return null;
  }

  const synthesisPath = path.join(projectDir, 'synthesis.md');
  fs.writeFileSync(synthesisPath, output, 'utf8');

  _runRetrospective(projectDir);

  return synthesisPath;
}

module.exports = {
  synthesize,
  parseRecommendation,
  _buildFindingsCorpus,
  _readDoctrine,
};
